README
# How to build a new pbtk model from .model files:
#First, create .model file and type in command prompt (with MCSIM/mod in path): mod -R <name>.model <name>.c
#This generates the <name>.c and <name>_inits.R files.
#In order to avoid duplicate names, function names in these newly created files must be changed.
#The changes (adding _<name>, an arbitrary choice) in the initialization file include: initparms_<name>, getParms_<name> (calling function from .c within initparms argument), Outputs_<name>, and initState_<name>.
#The changes in the .c file include: initmod_<name>, initforc_<name> (or delete function), getParms_<name>, derivs_<name>, jac_<name>, event_<name>, and root_<name>.
#These function names then must also be changed in the corresponding solve_pbtk function, specifically the state and parameter initilaizers as well as the function calls in the function ode.
#Parameter names passed to initparams must match those in .model file.

NEWS
News for R package httk

Changes in version 2.0.1 (February 28, 2020)
  New function set_httk_precision is now used throughout code to enforce a standard set of significant figures (4) and precision (nothing less than 1e-9).
  Added calc_hepatic_clearance wrapper function for calc_hep_clearance to allow backwards compatibility
  Fixed output of calc_mc_oral_equivalent (was sometimes returning all samples unasked, thank you Dan Dawson)
  Revised get_chemid to not crash in certain cases (thank you, Shannon Bell)

Changes in version 2.0.0 (February 10, 2020)
 	New generic inhalation PBPK model included, consistent with Linakis et al. (submitted) "Development and Evaluation of a High Throughput Inhalation Model for Organic Chemicals"
  New chemical specific parameters for volatile chemicals have been added:
    43 in human
    41 in rat
  Rewrote underlying code to allow more easy integration of new models. (goodbye, spaghetti code!)
  Rewritten funtions include:
    calc_analytic_css
    calc_mc_css
    convert_httkpop (renamed from convert_httk)
    solve_* model functions
    get_cheminfo
  Renamed a few httk-pop functions for clarity:
    httkpop_biotophys_default replaces httkpop_bio
    convert_httkpop replaces convert_httk
  New functions introduced:
    solve_model (mostly used by solve_* model functions)
    calc_mc_tk (performs Monte Carlo simulation using a solve_* function)
  Models must be much more thoroughly described now, with all relevant information placed in modelinfo_* files in the /R directory.
  New model-specific functions introduced:
    analytic_css_*: Model-specific analytic steady-state solution
    convert_httkpop_*: Model-specific functions for converting HTTK-pop 
      biometrics to model parameters
  EPA's DSSTox Chmeical Structure ID's (DTXSIDs, see http://comptox.epa.gov/dashboard/) now work as chemical identifiers in addition to name and CAS.
  Results now truncated to appropriate significant figures (4) and precision (1e-12).
  Beta testing and bug reports provided by Xiaoqing Chang.
  New physiological parameters have been added for monkeys
  To decrease package size the load image option of load_sipes2017 was eliminated
  Added vignette for Figure 6 from Frank, Christopher L., et al. "Defining toxicological tipping points in neuronal network development." Toxicology and Applied Pharmacology 354 (2018): 81-93.
   
Changes in version 1.10.1 (Septmeber 9, 2019)
  Bug fixes (thank you David Trudel).
  Changed all file name starting letters to lowercase.
  
Changes in version 1.10.0 (July 9, 2019)

  This version is consistent with the submitted manuscript Wambaugh et al. 
    "Assessing Toxicokinetic Uncertainty and Variability in Risk Prioritization"
  New human experimental measurements of fup and Clint are reported for 418 and 467 chemicals, respectively.
  Data on both fup and Clint are jointly available for 389 of those chemicals.
  Clint and fup values can now be either numeric values (as before) or distributions characterized by as "MEDIAN,LOWER95TH,UPPER95TH,PVALUE" for Clint and "MEDIAN,LOWER95TH,UPPER95TH" for fup. The code has been substantially revised to accomodate this.
  Added a minimum.Funbound.plasma since some of the Bayesian estimates are very low and at some point the values seem implausible. A value of 0.0001 was selected since it half the lowest reported measured value. Setting minimum.Funbound.plasma=0 removes this restriction.
  Monte Carlo coefficient of variation for Clint and fup has been divided into separate values for uncertainty (from measurement) and variability (population/genetic). Default values for coefficients of variation are fup.meas.cv=0.4, 
clint.meas.cv=0.3, fup.pop.cv=0.3, clint.pop.cv=0.3, (from Wambaugh et al, submitted). Note that most of the new fup mesurements have a lower CV than 0.3.
  All documentation converted to roxygen2 format.
  Vignette names have been updated to make the related publication clear.
  All references to "fub" have been converted to "fup" where appropriate.
  Rewrote calc_analytic_css to handle all models in the same manner.
  Corrected error where non-human species were using the incorrect p-value for Clint when default.to.human=TRUE (human p-value is now used). (thank you Jason Phillips and Shyam Patel for bug report).
  Changed arguments "mg" and "mol" for output.units in calc_mc_oral_equivalent to "mgpkgpday" and "umolpkgpday". (idea from Katie Paul-Friedman)
  Shyam Patel (Sciome) identified an error in how flow means were scaled by age in httk-pop Monte Carlo sampler.
  Changed httk-pop argument "fup.censor" to "fup.censored.dist".
  Armitage et al. (2014) model functions now work with input of vectors (to work well with data.table) or input of data.table
  Added the physchem parameters needed to run Armitage model
  Updated honda.ivive, reduced to 4 options as in Honda et al. (2019) Figure 8 panels a-d, changed "plasma.binding" to "bioactive.free.invivo", and exported function to allow user to call help file
  Added concentration as an option set by honda.ivive
  Added concentration = "tissue" as an option to calc_css functions
  Added bioactive.free.invivo as an option to calc_analytic_css functions, and calc_mc functions
  Function get_physchem_param: exported and now works with vectors of CAS and/or parameters
  Fixed calc_mc_css warnings
  
Changes in version 1.9.2 (April 22, 2019)

  Updated tests to reflect correct model predictions.
  Fixed errors that was causing the 3compartmentss and 1compartment models to not work with Monte Carlo. (thank you Johanna Nyffeler for bug report).

Changes in version 1.9.1 (April 15, 2019)

  Fixed signficant errors in calc_analytic_css that was causing Css to be over-estimated roughly 10x, therefore reducing the oral equivalent dose 10x (thank you Nisha Sipes for bug report).
   
Changes in version 1.9 (February 4, 2019)


src/gas.c
/* gasalation.c for R deSolve package
   ___________________________________________________

   Model File:  gasalation.model

   Date:  Mon Oct 22 11:15:43 2018

   Created by:  "C:/Users/mlinakis/Desktop/RLibrary/MCSIMU~1/mod/mod.exe v5.6.5"
    -- a model preprocessor by Don Maszle
   ___________________________________________________

   Copyright (c) 1993-2015 Free Software Foundation, Inc.

   Model calculations for compartmental model:

   14 States:
     Agutlumen = 0.0,
     Agut = 0.0,
     Aliver = 0.0,
     Aven = 0.0,
     Alung = 0.0,
     Aart = 0.0,
     Arest = 0.0,
     Akidney = 0.0,
     Atubules = 0.0,
     Ametabolized = 0.0,
     AUC = 0.0,
     Agas = 0.0,
     Aexh = 0.0,
     Amuc = 0.0,

   13 Outputs:
    "Cgut",
    "Cliver",
    "Cven",
    "Clung",
    "Cart",
    "Crest",
    "Ckidney",
    "Cplasma",
    "Aplasma",
    "Calv",
    "Cendexh",
    "Cmixexh",
    "Cmuc",

   1 Input:
     Cgas (forcing function)

   51 Parameters:
     BW = 0,
     Clmetabolismc = 0,
     vmax = 0,
     km = 0,
     hematocrit = 0,
     kgutabs = 0,
     Kkidney2pu = 0,
     Kliver2pu = 0,
     Krest2pu = 0,
     Kgut2pu = 0,
     Klung2pu = 0,
     Qcardiacc = 0,
     Qgfrc = 0,
     Qgutf = 0,
     Qkidneyf = 0,
     Qliverf = 0,
     Qlungf = 0,
     Vartc = 0,
     Vgutc = 0,
     Vkidneyc = 0,
     Vliverc = 0,
     Vlungc = 0,
     Vrestc = 0,
     Vvenc = 0,
     Fraction_unbound_plasma = 0,
     Rblood2plasma = 0,
     Clmetabolism = 0,
     Qcardiac = 0,
     Qgfr = 0,
     Qgut = 0,
     Qkidney = 0,
     Qliver = 0,
     Qlung = 0,
     Qrest = 0,
     Vart = 0,
     Vgut = 0,
     Vkidney = 0,
     Vliver = 0,
     Vlung = 0,
     Vrest = 0,
     Vven = 0,
     Qalvc = 0,
     Qalv = 0,
     Kblood2air = 0,
     kUrtc = 0,
     kUrt = 0,
     Kmuc2air = 0,
     Vmucc = 0,
     Vmuc = 0.0,
     Vmax = 0,
     Km = 0,
*/

#include <R.h>

/* Model variables: States */
#define ID_Agutlumen 0x00000
#define ID_Agut 0x00001
#define ID_Aliver 0x00002
#define ID_Aven 0x00003
#define ID_Alung 0x00004
#define ID_Aart 0x00005
#define ID_Arest 0x00006
#define ID_Akidney 0x00007
#define ID_Atubules 0x00008
#define ID_Ametabolized 0x00009
#define ID_AUC 0x0000a
#define ID_Agas 0x0000b
#define ID_Aexh 0x0000c
#define ID_Amuc 0x0000d

/* Model variables: Outputs */
#define ID_Cgut 0x00000
#define ID_Cliver 0x00001
#define ID_Cven 0x00002
#define ID_Clung 0x00003
#define ID_Cart 0x00004
#define ID_Crest 0x00005
#define ID_Ckidney 0x00006
#define ID_Cplasma 0x00007
#define ID_Aplasma 0x00008
#define ID_Calv 0x00009
#define ID_Cendexh 0x0000a
#define ID_Cmixexh 0x0000b
#define ID_Cmuc 0x0000c

/* Parameters */
static double parms[51];

#define BW parms[0]
#define Clmetabolismc parms[1]
#define vmax parms[2]
#define km parms[3]
#define hematocrit parms[4]
#define kgutabs parms[5]
#define Kkidney2pu parms[6]
#define Kliver2pu parms[7]
#define Krest2pu parms[8]
#define Kgut2pu parms[9]
#define Klung2pu parms[10]
#define Qcardiacc parms[11]
#define Qgfrc parms[12]
#define Qgutf parms[13]
#define Qkidneyf parms[14]
#define Qliverf parms[15]
#define Qlungf parms[16]
#define Vartc parms[17]
#define Vgutc parms[18]
#define Vkidneyc parms[19]
#define Vliverc parms[20]
#define Vlungc parms[21]
#define Vrestc parms[22]
#define Vvenc parms[23]
#define Fraction_unbound_plasma parms[24]
#define Rblood2plasma parms[25]
#define Clmetabolism parms[26]
#define Qcardiac parms[27]
#define Qgfr parms[28]
#define Qgut parms[29]
#define Qkidney parms[30]
#define Qliver parms[31]
#define Qlung parms[32]
#define Qrest parms[33]
#define Vart parms[34]
#define Vgut parms[35]
#define Vkidney parms[36]
#define Vliver parms[37]
#define Vlung parms[38]
#define Vrest parms[39]
#define Vven parms[40]
#define Qalvc parms[41]
#define Qalv parms[42]
#define Kblood2air parms[43]
#define kUrtc parms[44]
#define kUrt parms[45]
#define Kmuc2air parms[46]
#define Vmucc parms[47]
#define Vmuc parms[48]
#define Vmax parms[49]
#define Km parms[50]

/* Forcing (Input) functions */
static double forc[1];

#define Cgas forc[0]

/*----- Initializers */
void initmod_gas (void (* odeparms)(int *, double *))
{
  int N=51;
  odeparms(&N, parms);
}

void initforc_gas (void (* odeforcs)(int *, double *))
{
  int N=1;
  odeforcs(&N, forc);
}


void getParms_gas (double *inParms, double *out, int *nout) {
/*----- Model scaling */

  int i;

  for (i = 0; i < *nout; i++) {
    parms[i] = inParms[i];
  }


  kgutabs = kgutabs * 24 ;
  Clmetabolism = Clmetabolismc * 24 * BW ;
  Qcardiac = Qcardiacc * 24 * pow ( BW , 0.75 ) ;
  Qgfr = Qgfrc * pow ( BW , 0.75 ) * 24 ;
  Qgut = Qcardiac * Qgutf ;
  Qkidney = Qcardiac * Qkidneyf ;
  Qliver = Qcardiac * Qliverf ;
  Qlung = Qcardiac * Qlungf ;
  Qrest = Qcardiac - ( Qgut + Qkidney + Qliver + Qlung ) ;
  Vart = Vartc * BW ;
  Vgut = Vgutc * BW ;
  Vkidney = Vkidneyc * BW ;
  Vliver = Vliverc * BW ;
  Vlung = Vlungc * BW ;
  Vrest = Vrestc * BW ;
  Vven = Vvenc * BW ;
  Qalv = Qalvc * 24 * pow ( BW , 0.75 ) ;
  kUrt = kUrtc * pow ( BW , 0.75 ) * 24 ;
  Vmuc = Vmucc * BW ;
  Vmax = vmax * 60 * 24 ;
  Km = km ;

  for (i = 0; i < *nout; i++) {
    out[i] = parms[i];
  }
  }
/*----- Dynamics section */

void derivs_gas (int *neq, double *pdTime, double *y, double *ydot, double *yout, int *ip)
{

  yout[ID_Cgut] = y[ID_Agut] / Vgut ;

  yout[ID_Cliver] = y[ID_Aliver] / Vliver ;

  yout[ID_Cven] = y[ID_Aven] / Vven ;

  yout[ID_Clung] = y[ID_Alung] / Vlung ;

  yout[ID_Cart] = y[ID_Aart] / Vart ;

  yout[ID_Crest] = y[ID_Arest] / Vrest ;

  yout[ID_Ckidney] = y[ID_Akidney] / Vkidney ;

  yout[ID_Cplasma] = y[ID_Aven] / Vven / Rblood2plasma ;

  yout[ID_Aplasma] = y[ID_Aven] / Rblood2plasma * ( 1 - hematocrit ) ;

  yout[ID_Calv] = yout[ID_Cart] / Kblood2air ;

  yout[ID_Cendexh] = ( ( Qalv * yout[ID_Calv] ) + kUrt * ( ( yout[ID_Cmuc] / Kmuc2air ) - yout[ID_Calv] ) ) / Qalv ;

  yout[ID_Cmixexh] = 0.7 * yout[ID_Cendexh] + 0.3 * Cgas ;

  yout[ID_Cmuc] = y[ID_Amuc] / Vmuc ;

  ydot[ID_Agutlumen] = - kgutabs * y[ID_Agutlumen] ;

  ydot[ID_Agut] = kgutabs * y[ID_Agutlumen] + Qgut * ( yout[ID_Cart] - yout[ID_Cgut] * Rblood2plasma / Kgut2pu / Fraction_unbound_plasma ) ;

  ydot[ID_Aliver] = Qliver * yout[ID_Cart] + Qgut * yout[ID_Cgut] * Rblood2plasma / Kgut2pu / Fraction_unbound_plasma - ( Qliver + Qgut ) * yout[ID_Cliver] / Kliver2pu / Fraction_unbound_plasma * Rblood2plasma - Clmetabolism * yout[ID_Cliver] / Kliver2pu / Fraction_unbound_plasma * Rblood2plasma - Vmax * yout[ID_Cliver] / Kliver2pu / ( Km + yout[ID_Cliver] / Kliver2pu ) ;

  ydot[ID_Aven] = ( ( Qliver + Qgut ) * yout[ID_Cliver] / Kliver2pu + Qkidney * yout[ID_Ckidney] / Kkidney2pu + Qrest * yout[ID_Crest] / Krest2pu + Qlung * yout[ID_Clung] / Klung2pu ) * Rblood2plasma / Fraction_unbound_plasma - Qcardiac * yout[ID_Cven] ;

  ydot[ID_Alung] = Qlung * ( yout[ID_Cart] - yout[ID_Clung] * Rblood2plasma / Klung2pu / Fraction_unbound_plasma ) ;

  ydot[ID_Aart] = ( Qcardiac * ( yout[ID_Cven] - yout[ID_Cart] ) ) + ( Qalv * ( Cgas - yout[ID_Calv] ) ) - ( kUrt * ( Cgas - ( yout[ID_Cmuc] / Kmuc2air ) ) ) ;

  ydot[ID_Arest] = Qrest * ( yout[ID_Cart] - yout[ID_Crest] * Rblood2plasma / Krest2pu / Fraction_unbound_plasma ) ;

  ydot[ID_Akidney] = Qkidney * yout[ID_Cart] - Qkidney * yout[ID_Ckidney] / Kkidney2pu * Rblood2plasma / Fraction_unbound_plasma - Qgfr * yout[ID_Ckidney] / Kkidney2pu ;

  ydot[ID_Atubules] = Qgfr * yout[ID_Ckidney] / Kkidney2pu ;

  ydot[ID_Ametabolized] = Clmetabolism * yout[ID_Cliver] / Kliver2pu / Fraction_unbound_plasma * Rblood2plasma + Vmax * yout[ID_Cliver] / Kliver2pu / ( Km + yout[ID_Cliver] / Kliver2pu ) ;

  ydot[ID_AUC] = yout[ID_Cven] / Rblood2plasma ;

  ydot[ID_Agas] = ( Qalv * ( yout[ID_Calv] - Cgas ) ) + kUrt * ( ( yout[ID_Cmuc] / Kmuc2air ) - yout[ID_Calv] ) ;

  ydot[ID_Aexh] = ( Qalv * yout[ID_Calv] ) + kUrt * ( ( yout[ID_Cmuc] / Kmuc2air ) - yout[ID_Calv] ) ;

  ydot[ID_Amuc] = ( kUrt * ( Cgas - ( yout[ID_Cmuc] / Kmuc2air ) ) ) - ( kUrt * ( ( yout[ID_Cmuc] / Kmuc2air ) - yout[ID_Calv] ) ) ;

} /* derivs */


/*----- Jacobian calculations: */
void jac_gas (int *neq, double *t, double *y, int *ml, int *mu, double *pd, int *nrowpd, double *yout, int *ip)
{

} /* jac */


/*----- Events calculations: */
void event_gas (int *n, double *t, double *y)
{

} /* event */

/*----- Roots calculations: */
void root_gas (int *neq, double *t, double *y, int *ng, double *gout, double *out, int *ip)
{

} /* root */

